File: Scripts/eval-extraction.swift
#!/usr/bin/env swift
// eval-extraction.swift
//
// Runs the Qwen3-VL wine extraction prompt against every image in resources/
// and reports per-photo and aggregate results.
//
// Usage:
//   swift Scripts/eval-extraction.swift
//   swift Scripts/eval-extraction.swift --verbose   # prints all extracted wines

import Foundation

// MARK: - CLI

let verbose = CommandLine.arguments.contains("--verbose")
let ollamaBase = "http://localhost:11434"
let model = "qwen3-vl:8b"

// MARK: - Prompt (mirror of WineExtractionPrompt.swift)

let extractionPrompt = """
You are analyzing a photo of a restaurant wine list. Extract every wine you can identify.

Output exactly one JSON object per line (JSONL format). No surrounding array. No markdown. No explanation.

Each line must be valid JSON with this exact schema:
{"name":"...","producer":"...","vintage":"...","variety":"...","appellation":"...","price":"...","description":"...","listSection":"...","rawText":"...","confidence":0.95}

Field rules:
- name: wine name as printed on the list (required)
- producer: winery or producer name; often the same as name (null if unclear)
- vintage: 4-digit year as a string, e.g. "2019" (null if not shown)
- variety: grape variety or blend, e.g. "Cabernet Sauvignon" (null if not shown)
- appellation: region or appellation, e.g. "Napa Valley, California" (null if not shown)
- price: price as printed including currency symbol, e.g. "$48" or "48" (null if not shown)
- description: any tasting notes or description from the list (null if none)
- listSection: the section header under which this wine appears, e.g. "Red Wines" or "By the Glass" (null if no section header)
- rawText: the complete original text for this wine entry as it appears on the list
- confidence: float from 0.0 to 1.0 — your certainty that name and vintage are correctly extracted. Use 0.9+ for clear text, 0.6-0.8 for partially legible text, below 0.6 for guesses.

Output ONLY JSON lines. One wine per line. No other text.
"""

// MARK: - Types

struct WineObject: Decodable {
    let name: String
    let producer: String?
    let vintage: String?
    let variety: String?
    let appellation: String?
    let price: String?
    let description: String?
    let listSection: String?
    let rawText: String?
    let confidence: Double?
}

struct PhotoResult {
    let filename: String
    let wines: [WineObject]
    let parseErrors: [(line: String, error: String)]
    let httpStatus: Int?
    let networkError: String?
    let rawLineCount: Int
    let durationSeconds: Double
}

// MARK: - Helpers

func imageData(for path: String) -> Data? {
    guard let data = try? Data(contentsOf: URL(fileURLWithPath: path)) else { return nil }
    return data
}

func mimeType(for path: String) -> String {
    switch (path as NSString).pathExtension.lowercased() {
    case "jpg", "jpeg": return "image/jpeg"
    case "png":         return "image/png"
    case "webp":        return "image/webp"
    default:            return "image/jpeg"
    }
}

// MARK: - Ollama request

func toJpegData(path: String) -> Data? {
    let ext = (path as NSString).pathExtension.lowercased()
    if ["jpg", "jpeg", "png"].contains(ext) {
        return try? Data(contentsOf: URL(fileURLWithPath: path))
    }
    // Convert webp (and other formats) to JPEG via sips, capped at 1500px to stay under Ollama's request size limit
    let tmp = URL(fileURLWithPath: NSTemporaryDirectory())
        .appendingPathComponent(UUID().uuidString + ".jpg").path
    let proc = Process()
    proc.executableURL = URL(fileURLWithPath: "/usr/bin/sips")
    proc.arguments = ["-s", "format", "jpeg", "-Z", "1500", path, "--out", tmp]
    proc.standardOutput = FileHandle.nullDevice
    proc.standardError = FileHandle.nullDevice
    try? proc.run(); proc.waitUntilExit()
    let data = try? Data(contentsOf: URL(fileURLWithPath: tmp))
    try? FileManager.default.removeItem(atPath: tmp)
    return data
}

func extractWines(from imagePath: String) -> PhotoResult {
    let filename = (imagePath as NSString).lastPathComponent
    let start = Date()

    guard let imgData = toJpegData(path: imagePath) else {
        return PhotoResult(filename: filename, wines: [], parseErrors: [],
                          httpStatus: nil, networkError: "could not read/convert file",
                          rawLineCount: 0, durationSeconds: 0)
    }

    let b64 = imgData.base64EncodedString()
    // /api/chat with assistant pre-fill of "{" bypasses Qwen3-VL thinking mode.
    let body: [String: Any] = [
        "model": model,
        "messages": [
            ["role": "user", "content": extractionPrompt, "images": [b64]],
            ["role": "assistant", "content": "{"]
        ],
        "stream": false,
        "options": ["temperature": 0.1]
    ]

    guard let url = URL(string: "\(ollamaBase)/api/chat"),
          let bodyData = try? JSONSerialization.data(withJSONObject: body) else {
        return PhotoResult(filename: filename, wines: [], parseErrors: [],
                          httpStatus: nil, networkError: "could not build request",
                          rawLineCount: 0, durationSeconds: 0)
    }

    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.httpBody = bodyData
    request.timeoutInterval = 180

    var responseData: Data?
    var httpStatus: Int?
    var networkErrorMsg: String?

    let sem = DispatchSemaphore(value: 0)
    URLSession.shared.dataTask(with: request) { data, response, error in
        httpStatus = (response as? HTTPURLResponse)?.statusCode
        responseData = data
        networkErrorMsg = error?.localizedDescription
        sem.signal()
    }.resume()
    sem.wait()

    let duration = Date().timeIntervalSince(start)

    if let err = networkErrorMsg {
        return PhotoResult(filename: filename, wines: [], parseErrors: [],
                          httpStatus: httpStatus, networkError: err,
                          rawLineCount: 0, durationSeconds: duration)
    }

    guard httpStatus == 200, let responseData else {
        return PhotoResult(filename: filename, wines: [], parseErrors: [],
                          httpStatus: httpStatus, networkError: "HTTP \(httpStatus ?? -1)",
                          rawLineCount: 0, durationSeconds: duration)
    }

    // /api/chat non-stream: message.content holds the response text.
    // Prepend "{" to restore the pre-filled opening brace.
    guard let envelope = try? JSONSerialization.jsonObject(with: responseData) as? [String: Any],
          let message = envelope["message"] as? [String: Any],
          let content = message["content"] as? String else {
        let preview = String(data: responseData, encoding: .utf8).map { String($0.prefix(200)) } ?? "(binary)"
        return PhotoResult(filename: filename, wines: [], parseErrors: [("(envelope)", preview)],
                          httpStatus: 200, networkError: nil,
                          rawLineCount: 0, durationSeconds: duration)
    }
    let rawText = "{" + content

    // Parse JSONL — one wine per line
    let lines = rawText
        .components(separatedBy: .newlines)
        .map { $0.trimmingCharacters(in: .whitespaces) }
        .filter { !$0.isEmpty }

    var wines: [WineObject] = []
    var parseErrors: [(line: String, error: String)] = []

    for line in lines {
        guard let lineData = line.data(using: .utf8) else { continue }
        do {
            let wine = try JSONDecoder().decode(WineObject.self, from: lineData)
            wines.append(wine)
        } catch {
            // Skip lines that are clearly not JSON objects (e.g. thinking text)
            if line.hasPrefix("{") {
                parseErrors.append((line: String(line.prefix(120)), error: error.localizedDescription))
            }
        }
    }

    return PhotoResult(filename: filename, wines: wines, parseErrors: parseErrors,
                      httpStatus: 200, networkError: nil,
                      rawLineCount: lines.count, durationSeconds: duration)
}

// MARK: - Discover photos

let resourcesDir = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
    .appendingPathComponent("resources")

let extensions = Set(["jpg", "jpeg", "png", "webp"])
let photoPaths: [String]

do {
    let items = try FileManager.default.contentsOfDirectory(at: resourcesDir,
        includingPropertiesForKeys: nil)
    photoPaths = items
        .filter { extensions.contains($0.pathExtension.lowercased()) }
        .map(\.path)
        .sorted()
} catch {
    print("ERROR: could not read resources/ directory: \(error)")
    exit(1)
}

guard !photoPaths.isEmpty else {
    print("ERROR: no image files found in resources/")
    exit(1)
}

// MARK: - Verify Ollama is reachable

let pingURL = URL(string: "\(ollamaBase)/api/tags")!
let pingDone = DispatchSemaphore(value: 0)
var ollamaReachable = false
URLSession.shared.dataTask(with: pingURL) { data, resp, _ in
    ollamaReachable = (resp as? HTTPURLResponse)?.statusCode == 200
    pingDone.signal()
}.resume()
pingDone.wait()

guard ollamaReachable else {
    print("ERROR: Ollama not reachable at \(ollamaBase)")
    print("Start it with: ollama serve")
    exit(1)
}

// MARK: - Run eval

print("Running extraction eval against \(photoPaths.count) photos using \(model)")
print("Each inference takes ~10–60s depending on hardware.\n")

var allResults: [PhotoResult] = []

for (idx, path) in photoPaths.enumerated() {
    let filename = (path as NSString).lastPathComponent
    print("[\(idx+1)/\(photoPaths.count)] \(filename)…", terminator: " ")
    fflush(stdout)

    let result = extractWines(from: path)
    allResults.append(result)

    if let err = result.networkError {
        print("ERROR — \(err)")
    } else {
        let lowConf = result.wines.filter { ($0.confidence ?? 1.0) < 0.7 }.count
        let confStr = lowConf > 0 ? " (\(lowConf) low-conf)" : ""
        let parseStr = result.parseErrors.isEmpty ? "" : " \(result.parseErrors.count) parse error(s)"
        print("\(result.wines.count) wines\(confStr)\(parseStr) in \(String(format: "%.1f", result.durationSeconds))s")
    }

    if verbose {
        for wine in result.wines {
            let conf = wine.confidence.map { String(format: "%.2f", $0) } ?? "?"
            let vintage = wine.vintage ?? "NV"
            let section = wine.listSection.map { " [\($0)]" } ?? ""
            print("    \(conf) \(wine.name) \(vintage)\(section)")
        }
        for (line, err) in result.parseErrors {
            print("    PARSE ERROR: \(err)")
            print("    LINE: \(line)")
        }
    }
}

// MARK: - Summary

print("\n━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
print("EXTRACTION EVAL REPORT — \(model)")
print("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n")

let successful = allResults.filter { $0.networkError == nil && $0.httpStatus == 200 }
let totalWines = successful.flatMap { $0.wines }.count
let totalParseErrors = successful.flatMap { $0.parseErrors }.count
let allConfidences = successful.flatMap { $0.wines }.compactMap { $0.confidence }
let avgConf = allConfidences.isEmpty ? 0.0 : allConfidences.reduce(0, +) / Double(allConfidences.count)
let lowConfCount = allConfidences.filter { $0 < 0.7 }.count
let avgDuration = successful.isEmpty ? 0.0 : successful.map { $0.durationSeconds }.reduce(0, +) / Double(successful.count)

print("Per-photo breakdown:")
print(String(repeating: "─", count: 76))

let maxName = allResults.map { $0.filename.count }.max() ?? 20
for r in allResults {
    let name = r.filename.padding(toLength: max(maxName, 20), withPad: " ", startingAt: 0)
    if let err = r.networkError {
        print("  \(name)  ERROR: \(err)")
    } else {
        let wines = String(r.wines.count).padding(toLength: 3, withPad: " ", startingAt: 0)
        let parseE = r.parseErrors.isEmpty ? "  " : "\(r.parseErrors.count)!"
        let lowC = r.wines.filter { ($0.confidence ?? 1.0) < 0.7 }.count
        let lowStr = lowC > 0 ? " \(lowC) low-conf" : ""
        let dur = String(format: "%.1fs", r.durationSeconds)
        print("  \(name)  \(wines) wines  \(parseE)  \(dur)\(lowStr)")
    }
}

print(String(repeating: "─", count: 76))
print()
print("Total wines extracted:   \(totalWines) across \(successful.count)/\(allResults.count) photos")
print("Avg wines per photo:     \(successful.isEmpty ? 0 : totalWines / successful.count)")
print("Parse errors:            \(totalParseErrors)")
print("Avg confidence:          \(String(format: "%.2f", avgConf))")
print("Low-confidence (<0.7):   \(lowConfCount)/\(allConfidences.count) (\(allConfidences.isEmpty ? 0 : Int(Double(lowConfCount)/Double(allConfidences.count)*100))%)")
print("Avg inference time:      \(String(format: "%.1f", avgDuration))s per photo")

// Section header coverage
let winesWithSection = successful.flatMap { $0.wines }.filter { $0.listSection != nil }.count
print("Section header captured: \(winesWithSection)/\(totalWines) wines (\(totalWines == 0 ? 0 : Int(Double(winesWithSection)/Double(totalWines)*100))%)")

// Vintage coverage
let winesWithVintage = successful.flatMap { $0.wines }.filter { $0.vintage != nil }.count
print("Vintage captured:        \(winesWithVintage)/\(totalWines) wines (\(totalWines == 0 ? 0 : Int(Double(winesWithVintage)/Double(totalWines)*100))%)")

// Price coverage
let winesWithPrice = successful.flatMap { $0.wines }.filter { $0.price != nil }.count
print("Price captured:          \(winesWithPrice)/\(totalWines) wines (\(totalWines == 0 ? 0 : Int(Double(winesWithPrice)/Double(totalWines)*100))%)")

print()
if totalParseErrors == 0 && avgConf >= 0.8 && lowConfCount == 0 {
    print("VERDICT: ✓ PASS — prompt is production-ready")
} else if avgConf >= 0.7 && Double(totalParseErrors) / max(Double(totalWines), 1) < 0.05 {
    print("VERDICT: ~ MARGINAL — acceptable but review low-confidence wines; run --verbose to inspect")
} else {
    print("VERDICT: ✗ FAIL — iterate on WineExtractionPrompt.swift before shipping")
    if totalParseErrors > 0 {
        print("  → \(totalParseErrors) lines failed to parse as JSON — model may be adding commentary")
    }
    if avgConf < 0.7 {
        print("  → avg confidence \(String(format: "%.2f", avgConf)) — model is uncertain; check dim/angled photos")
    }
}
print("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n")

File: backend/tests/eval_extraction.py
"""
Integration test: extraction quality eval against resources/images/.

Sends each photo in resources/images/ to the backend's /scan endpoint,
parses the SSE stream, and asserts quality metrics across the corpus.
Mirrors the logic of Scripts/eval-extraction.swift but tests the full
backend API surface rather than calling Ollama directly.

Run with:
    cd backend
    uv sync --extra dev
    BACKEND_URL=http://localhost:8000 pytest -m integration tests/eval_extraction.py -v -s

Requires:
    - Backend running at BACKEND_URL (default http://localhost:8000)
    - Ollama running with qwen3-vl:8b and the backend configured to reach it
    - At least one image in resources/images/ (see resources/README.md)
"""

import json
import os
import time
from pathlib import Path

import httpx
import pytest

BACKEND_URL = os.environ.get("BACKEND_URL", "http://localhost:8000")
RESOURCES_DIR = Path(__file__).parent.parent.parent / "resources" / "images"
IMAGE_EXTENSIONS = {".jpg", ".jpeg", ".png", ".webp"}

MIN_AVG_CONFIDENCE = 0.70
MAX_ERROR_EVENT_RATE = 0.05
MIN_WINES_PER_PHOTO = 1


def _parse_sse_events(text: str) -> list[dict[str, object]]:
    """Parse an SSE response body into a list of {event, data} dicts."""
    events: list[dict[str, object]] = []
    current_event: str | None = None
    current_data: list[str] = []

    for line in text.splitlines():
        if line.startswith("event: "):
            current_event = line[7:].strip()
        elif line.startswith("data: "):
            current_data.append(line[6:].strip())
        elif line == "" and current_event and current_data:
            try:
                parsed = json.loads(" ".join(current_data))
                events.append({"event": current_event, "data": parsed})
            except json.JSONDecodeError:
                pass
            current_event = None
            current_data = []

    return events


@pytest.fixture(scope="module")
def image_paths() -> list[Path]:
    if not RESOURCES_DIR.exists():
        pytest.skip(f"resources/images/ not found at {RESOURCES_DIR}")
    paths = sorted(p for p in RESOURCES_DIR.iterdir() if p.suffix.lower() in IMAGE_EXTENSIONS)
    if not paths:
        pytest.skip("No images found in resources/images/ — see resources/README.md")
    return paths


@pytest.mark.integration
def test_backend_reachable_for_eval() -> None:
    """Fail fast with a clear message if the backend is not running."""
    try:
        resp = httpx.get(f"{BACKEND_URL}/health", timeout=5)
        assert resp.status_code == 200, f"Health check returned {resp.status_code}"
    except httpx.ConnectError:
        pytest.fail(
            f"Backend not reachable at {BACKEND_URL}. "
            "Start it with: cd backend && uvicorn backend.main:app --reload"
        )


@pytest.mark.integration
def test_extraction_quality(image_paths: list[Path]) -> None:
    """
    End-to-end extraction quality eval: POST each image to /scan, parse SSE
    stream, and assert corpus-level quality metrics.

    Per-photo and aggregate metrics are printed to stdout (use -s to see them).
    """
    all_wines: list[dict[str, object]] = []
    error_event_count = 0
    per_photo: list[dict[str, object]] = []

    for img_path in image_paths:
        start = time.perf_counter()
        img_bytes = img_path.read_bytes()

        with httpx.stream(
            "POST",
            f"{BACKEND_URL}/scan",
            files={"photo": (img_path.name, img_bytes, "image/jpeg")},
            timeout=300,
        ) as response:
            assert response.status_code == 200, (
                f"{img_path.name}: /scan returned HTTP {response.status_code}"
            )
            body = response.read().decode("utf-8")

        duration = time.perf_counter() - start
        events = _parse_sse_events(body)

        wines = [e["data"] for e in events if e["event"] == "wine"]
        errors = [e["data"] for e in events if e["event"] == "error"]

        all_wines.extend(wines)  # type: ignore[arg-type]
        error_event_count += len(errors)

        per_photo.append(
            {
                "filename": img_path.name,
                "wine_count": len(wines),
                "error_count": len(errors),
                "duration": duration,
            }
        )

    # Print report
    confidences = [
        float(w["confidence"])  # type: ignore[arg-type]
        for w in all_wines
        if "confidence" in w  # type: ignore[operator]
    ]
    avg_conf = sum(confidences) / len(confidences) if confidences else 0.0
    low_conf = sum(1 for c in confidences if c < 0.7)
    error_rate = error_event_count / max(len(all_wines), 1)

    print(f"\n{'=' * 72}")
    print(f"EXTRACTION EVAL — {len(image_paths)} photo(s)  backend={BACKEND_URL}")
    print(f"{'=' * 72}")
    for r in per_photo:
        print(
            f"  {str(r['filename']):<42} {r['wine_count']:>3} wines  "
            f"{r['error_count']} errors  {r['duration']:.1f}s"
        )
    print(f"\nTotal wines extracted : {len(all_wines)}")
    print(f"Avg confidence        : {avg_conf:.2f}")
    print(f"Low-confidence (<0.7) : {low_conf}/{len(all_wines)}")
    print(f"SSE error events      : {error_event_count}")
    print(f"Error rate            : {error_rate:.2%}")
    print(f"{'=' * 72}")

    # Assertions
    assert len(all_wines) > 0, "No wines extracted from any image"
    assert avg_conf >= MIN_AVG_CONFIDENCE, (
        f"Avg confidence {avg_conf:.2f} below threshold {MIN_AVG_CONFIDENCE}"
    )
    assert error_rate <= MAX_ERROR_EVENT_RATE, (
        f"SSE error event rate {error_rate:.2%} exceeds threshold {MAX_ERROR_EVENT_RATE:.2%}"
    )
    for r in per_photo:
        assert int(str(r["wine_count"])) >= MIN_WINES_PER_PHOTO, (
            f"{r['filename']}: extracted 0 wines (expected ≥ {MIN_WINES_PER_PHOTO})"
        )
